assignments/tasks.go

```diff
@@ -0,0 +1,158 @@
+	if !bytes.HasPrefix(contents, []byte("# ")) {
+		return nil, fmt.Errorf("task with name: %s, does not start with a # title marker", name)
+	}
+	bodyIndex := bytes.Index(contents, []byte("\n\n"))
+	if bodyIndex == -1 {
+		return nil, fmt.Errorf("failed to find task body in task: %s", name)
+	}
+
+	return &pb.Task{
+		AssignmentOrder: assignmentOrder,
+		Title:           string(contents[2:bodyIndex]),
+		Body:            string(contents[bodyIndex+2:]),
+		Name:            name,
+	}, nil
+}
+
+// tasksFromAssignments returns a map, mapping each assignment-order to a map of tasks.
+func tasksFromAssignments(assignments []*pb.Assignment) map[uint32]map[string]*pb.Task {
+	taskMap := make(map[uint32]map[string]*pb.Task)
+	for _, assignment := range assignments {
+		temp := make(map[string]*pb.Task)
+		for _, task := range assignment.Tasks {
+			temp[task.Name] = task
+		}
+		taskMap[assignment.Order] = temp
+	}
+	return taskMap
+}
+
+// mapTasksByID returns a map, mapping each task to its ID.
+func mapTasksByID(tasks []*pb.Task) map[uint64]*pb.Task {
+	taskMap := make(map[uint64]*pb.Task)
+	for _, task := range tasks {
+		taskMap[task.ID] = task
+	}
+	return taskMap
+}
+
+func handleTasks(ctx context.Context, db database.Database, sc scm.SCM, course *pb.Course, assignments []*pb.Assignment) error {
+	createdIssues := []*pb.Issue{}
+	var err error
+
+	tasksFromTestsRepo := tasksFromAssignments(assignments)
+	createdTasks, updatedTasks, deletedTasks, err := db.SynchronizeAssignmentTasks(course, tasksFromTestsRepo)
+	if err != nil {
+		return err
+	}
+
+	repos, err := db.GetRepositoriesWithIssues(&pb.Repository{
+		OrganizationID: course.GetOrganizationID(),
+	})
+	if err != nil {
+		return err
+	}
+
+	// Creates, updates and deletes issues on all group repositories, based on how tasks differ from last push.
+	for _, repo := range repos {
+		if !repo.IsGroupRepo() {
+			continue
+		}
+
+		repoCreatedIssues, err := createIssues(ctx, sc, course, repo, createdTasks)
+		if err != nil {
+			return err
+		}
+		createdIssues = append(createdIssues, repoCreatedIssues...)
+		if err = updateIssues(ctx, sc, course, repo, updatedTasks, false); err != nil {
+			return err
+		}
+		if err = updateIssues(ctx, sc, course, repo, deletedTasks, true); err != nil {
+			return err
+		}
+	}
+
+	// Deleting issues from database that no longer has an associated task.
+	if err = db.DeleteIssuesOfAssociatedTasks(deletedTasks); err != nil {
+		return err
+	}
+	// Creating issues in database, based on issues created on scm.
+	return db.CreateIssues(createdIssues)
+}
+
+// createIssues creates issues on scm based on repository, course and tasks. Returns created issues.
+func createIssues(ctx context.Context, sc scm.SCM, course *pb.Course, repo *pb.Repository, tasks []*pb.Task) ([]*pb.Issue, error) {
+	createdIssues := []*pb.Issue{}
+	for _, task := range tasks {
+		issueOptions := &scm.CreateIssueOptions{
+			Organization: course.Name,
```

Comment:
Is this correct? The `course.Name` is the long name of the course, e.g., `Operating Systems and Systems Programming`, whereas `course.Tag` would typically be `DAT320` (capitalized). There is also `course.GetOrganizationPath()` that might be more accurate for this one...

Reply:
I don't know... IssueOptions was formatted like this when I started on this project. When running tests on the `qf101` org, `course.Name` is set to this value. Given what you said, I suppose `course.Tag` or `course.GetOrganizationPath()` should be used instead.

Reply:
According to the github api it should be as follows: `/repos/{owner}/{repo}/issues`, where `owner` is what `course.Name` sets atm

Reply:
Right, so `owner` in this context is the organization (like `dat320-2022`) or if a personal user (like `meling`), whereas `repo` is the various repos under said organization/user. Hence, this should be `course.GetOrganizationPath()`, which corresponds to this line in the proto file:

```proto
    string organizationPath        = 9;   // The organization's SCM name, e.g. uis-dat520-2020.
```